Tribler/community/tunnel/hidden_community.py

```diff
@@ -228,53 +258,103 @@ def monitor_downloads(self, dslist):
 
             blacklist = self.dht_blacklist[info_hash]
 
+            if self.notifier:
+                self.notifier.notify(NTFY_TUNNEL, NTFY_DHT_LOOKUP, info_hash.encode('hex')[:6], peers)
+
             # cleanup dht_blacklist
             for i in xrange(len(blacklist) - 1, -1, -1):
                 if time.time() - blacklist[i][0] > 60:
                     blacklist.pop(i)
-
             exclude = [rp[2] for rp in self.my_download_points.values()] + [sock_addr for _, sock_addr in blacklist]
-            for peer in set(peers):
+            for peer in peers:
                 if peer not in exclude:
+                    self.add_discovered_candidate(Candidate((peer[0], peer[1]), tunnel=False))
                     self._logger.debug("Requesting key from %s", peer)
-
                     # Blacklist this sock_addr for a period of at least 60s
                     self.dht_blacklist[info_hash].append((time.time(), peer))
-
                     self.create_key_request(info_hash, peer)
 
-        self._logger.debug("Doing dht lookup for hidden community")
-        self.last_dht_lookup[info_hash] = time.time()
-        self.dht_lookup(info_hash, dht_callback)
-
-    def create_key_request(self, info_hash, sock_addr):
-        # 1. Select a circuit
-        self._logger.debug("Create key request: select circuit")
-        circuit = self.selection_strategy.select(None, self.hops[info_hash])
-        if not circuit:
-            self._logger.error("No circuit for key-request")
+    def create_key_request(self, info_hash, sock_addr, max_retries=3):
+        self._logger.debug("Create key request: build a circuit to the intropoint")
+
+        def cb(circuit):
+            # Send a key-request message over the newly created circuit
+            if self.notifier:
+                self.notifier.notify(NTFY_TUNNEL, NTFY_KEY_REQUEST, info_hash.encode('hex')[:6], sock_addr)
+            self._logger.debug("Create key request: send key request")
+            cache = self.request_cache.add(KeyRequestCache(self, circuit, sock_addr, info_hash))
+            meta = self.get_meta_message(u'key-request')
+            message = meta.impl(distribution=(self.global_time,), payload=(circuit.circuit_id, cache.number, info_hash))
+            circuit.tunnel_data(sock_addr, TUNNEL_PREFIX + message.packet)
+
+        candidate = self.get_candidate(sock_addr, replace=False)
+        if not candidate or candidate.get_category(time.time()) not in (u"walk", u"stumble"):
+            if max_retries > 0:
+                self._logger.debug("Create key request: retry introducing candidate")
+                retry_lambda = lambda i = info_hash, s = sock_addr, r = max_retries - 1: self.create_key_request(i, s, r)
+                self.register_task(retry_lambda, reactor.callLater(5, retry_lambda))
             return False
 
-        # 2. Send a key-request message
-        self._logger.debug("Create key request: send key request")
```

Comment:
Similar to before, I wouldn't create a new circuit for a key-request. It doesn't make a lot of sense. Sock_addr is supposed to be connectable, hence we can use any circuit to send it a message. As long as the port received from the dht is the port where Dispersy is running.